mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m

```diff
@@ -0,0 +1,185 @@
+
+  if (!exists) {
+    NSError *createError = nil;
+    os_log_info([KMLogs dataLog], "createDataDirectoryIfNecessary, about to attempt createDirectoryAtPath for: '%{public}@'", self.keymanDataDirectory.path);
+    [fileManager createDirectoryAtPath:self.keymanDataDirectory.path withIntermediateDirectories:YES attributes:nil error:nil];
+    if (createError) {
+      os_log_error([KMLogs dataLog], "error creating Keyman data directory: '%{public}@'", createError.localizedDescription);
+    } else {
+      os_log_info([KMLogs dataLog], "created Keyman data directory: '%{public}@'", self.keymanDataDirectory.path);
+    }
+  } else {
+    os_log_info([KMLogs dataLog], "Keyman data directory already exists: '%{public}@'", self.keymanDataDirectory.path);
+  }
+}
+
+/**
+ * Creates Keyman keyboard directory if it does not exist yet. This is the 'Keyman-Keyboards' directory.
+ * It should not be created until after migrating because its existence would block migrating data from the old location.
+ */
+- (void)createKeyboardsDirectoryIfNecessary {
+  NSFileManager *fileManager = [NSFileManager defaultManager];
+  BOOL isDir;
+  BOOL exists = [fileManager fileExistsAtPath:self.keymanKeyboardsDirectory.path isDirectory:&isDir];
+
+  if (!exists) {
+    NSError *createError = nil;
+    os_log_info([KMLogs dataLog], "createKeyboardsDirectoryIfNecessary, about to attempt createDirectoryAtPath for: '%{public}@'", self.keymanKeyboardsDirectory.path);
+    [fileManager createDirectoryAtPath:self.keymanKeyboardsDirectory.path withIntermediateDirectories:YES attributes:nil error:nil];
+    if (createError) {
+      os_log_error([KMLogs dataLog], "error creating Keyman-Keyboards directory: '%{public}@'", createError.localizedDescription);
+    } else {
+      os_log_info([KMLogs dataLog], "created Keyman-Keyboards subdirectory: '%{public}@'", self.keymanKeyboardsDirectory.path);
+    }
+  } else {
+    os_log_info([KMLogs dataLog], "Keyman-Keyboards already exists: '%{public}@'", self.keymanKeyboardsDirectory.path);
+  }
+}
+
+- (NSURL *)obsoleteKeymanKeyboardsDirectory {
+  if (_obsoleteKeymanKeyboardsDirectory == nil) {
+    NSURL *keymanUrl = [self.documentsSubDirectory URLByAppendingPathComponent:kKeyboardsDirectoryName  isDirectory: TRUE];
+    _obsoleteKeymanKeyboardsDirectory = keymanUrl;
+  }
+  return _obsoleteKeymanKeyboardsDirectory;
+}
+
+- (BOOL)keyboardsExistInObsoleteDirectory {
+  NSFileManager *fileManager = [NSFileManager defaultManager];
+  BOOL isDir;
+  BOOL exists = ([fileManager fileExistsAtPath:self.obsoleteKeymanKeyboardsDirectory.path isDirectory:&isDir]);
+  return exists;
+}
+
+/**
+ * Migrate the keyboards data from the old location in '/Documents' to the new location '/Application Support/keyman.inputmethod.Keyman/'
+ */
+- (BOOL)migrateData {
+  BOOL didMoveData = NO;
+  NSFileManager *fileManager = [NSFileManager defaultManager];
+  BOOL dataExistsInOldLocation = [self keyboardsExistInObsoleteDirectory];
```

Comment:
Yes, I added a comment for this and another function called only by migrateData that would also trigger the access request. 

This would only happen if the Keyman was configured to store data in ~/Documents (so that the settings in UserDefaults were updated with the location) but then they revoked those privileges just before installing the new version of Keyman that would attempt the migration.